Mute error occurrences

crbelaus opened the pull request:
This pull request adds a new `muted` field to errors. Noisy errors can be muted in the UI so integrations and notifications can ignore them as well.

In the end I decided to use the _mute_ instead of _silence_ as I think that _mute_ and _unmute_ are fairly commonly understood terms.

This is an addition to the capability of ignoring errors that we already have, and enables the possibility of tracking error occurrences but don't notify them. Closes #117


New mute action in the error detail page:
<img width="1552" alt="New mute action in the error detail page" src="https://github.com/user-attachments/assets/11663fcc-e233-4746-b5c2-a8fa7e233e3a" />

Mute and unmute actions in the dashboard page
<img width="1552" alt="Mute and unmute actions in the dashboard page" src="https://github.com/user-attachments/assets/182f9add-e72c-44aa-9e6f-3769b4f88f71" />

crbelaus commented on test/error_tracker/telemetry_test.exs:
```diff
@@ -11,22 +11,36 @@ defmodule ErrorTracker.TelemetryTest do
 
-    refute_receive {:telemetry_event, [:error_tracker, :error, :new], _,
-                    %{occurrence: %Occurrence{}}},
+    refute_receive {:telemetry_event, [:error_tracker, :error, :new], _, %{error: %Error{}}},
```
This was wrong. We were checking the telemetry event incorrectly and it did never match because it has an `%Error{}` instead of an `%Occurrence{}`.

The reason is that every time we raised the error in a different line, so they were different errors and not the same.

The new code ensures that the same error is raised multiple times and helped me catch this error.

odarriba replied:
Awesome!

odarriba commented on lib/error_tracker.ex:
```diff
@@ -300,8 +328,16 @@ defmodule ErrorTracker do
+    {existing_status, muted} =
+      case Repo.one(status_and_muted_query) do
+        {existing_status, muted} -> {existing_status, muted}
+        nil -> {nil, nil}
```
It works as it is, but to keep type consistency it would be:
```suggestion
        nil -> {nil, false}
```

crbelaus replied:
Good catch! Fixed!

odarriba commented on lib/error_tracker/migration/sqlite/v05.ex:
```diff
@@ -0,0 +1,17 @@
+
+  def down(_opts) do
+    alter table(:error_tracker_errors) do
+      remove :muted
```
This will not work, as SQLite does not support removing columns. We can leave it as it is  or emit an exception, as your wish 😅

crbelaus replied:
So it looks like this was not supported in old SQLite versions but it has been [supported](https://sqlite.org/lang_altertable.html#altertabdropcol) for a while.

I just tried in my local environment (added a V06 SQLite migration that did this) and can confirm that it works fine now.

odarriba commented on lib/error_tracker/schemas/error.ex:
```diff
@@ -22,6 +22,7 @@ defmodule ErrorTracker.Error do
     field :status, Ecto.Enum, values: [:resolved, :unresolved], default: :unresolved
     field :fingerprint, :binary
     field :last_occurrence_at, :utc_datetime_usec
+    field :muted, :boolean, default: false
```
I would not set the default, rather than leaving it in the schema migration. 

It is a personal preference (and if you want to leave it as it is I'm not opposing to it), but with a default value you will see that getting a partial struct from a query with a `select` can cause to receivbe a `false`on this field that may not be the real value in DB.

Feel free to change it or leave it as it is, no problem :D

crbelaus replied:
Agreed. I've removed the default from the codebase and left it on the database only.

odarriba reviewed (approved):
Looks good to me! I added some comments but all of them minor.

My only worry is that I don't know if it is better to avoid sending telemetry events, or send them with a `muted: true` information.

crbelaus commented:
> My only worry is that I don't know if it is better to avoid sending telemetry events, or send them with a `muted: true` information.

This makes sense. I've updated the code so the `[:error_tracker, :occurrence, :new]` telemetry event is always emitted but it now also includes a `muted` metadata so users can ignore those events as needed. Since we are adding more metadata I've also added the `error` in case it may be useful.

I've also updated the documentation to reflect this changes.

odarriba reviewed (approved):
🚀 🤌
